File: python/bayesmixpy/build_bayesmix.py
import os
import pathlib
import subprocess
import sys

from distutils.spawn import find_executable

from dotenv import set_key

from .shell_utils import get_env_file, run_shell

HERE = os.path.dirname(os.path.realpath(__file__))
path = pathlib.Path(HERE)
BAYESMIX_HOME = os.environ.get("BAYESMIX_HOME", path.resolve().parents[1])

py2to3 = find_executable("2to3")
PROTO_DIR = os.path.join(path, "proto/")


def set_bayesmix_env(run_path):
    env_file = get_env_file()
    if not os.path.exists(env_file):
        open(env_file, mode='a').close()

    set_key(env_file, "BAYESMIX_EXE", run_path)


def build_bayesmix(nproc=1, build_dirname="build"):
    """
    Builds the BayesMix executable. After the build, if no error has occurred,
    it prints out the path to the executable. Save the path into the environment
    variable BAYESMIX_EXE.

    Parameters
    ----------

    nproc : int
        Number of processes to use for parallel compilation.
    """
    print("Building the Bayesmix executable")
    build_dir = os.path.join(BAYESMIX_HOME, build_dirname)
    os.makedirs(build_dir, exist_ok=True)
    cmake_cmd = "cmake .. -DDISABLE_BENCHMARKS=TRUE -DDISABLE_TESTS=TRUE " + \
        "-DDISABLE_PLOTS=TRUE -DCMAKE_BUILD_TYPE=Release"
    try:
        run_shell(cmake_cmd, cwd=build_dir)
    except subprocess.CalledProcessError as e:
        print(e)
        print("Some error has occurred while building Bayesmix. The library has not"
              " been installed!")
        return

    run_cmd = "make run_mcmc -j{}".format(nproc)
    try:
        run_shell(run_cmd, cwd=build_dir)
    except subprocess.CalledProcessError as e:
        print(e)
        print("Some error has occurred while building Bayesmix. The library has not"
              " been installed!")
        return

    set_bayesmix_env("{0}/{1}".format(build_dir, "run_mcmc"))

    two_to_three_command = [
            py2to3, "--output-dir={0}".format(PROTO_DIR), "-W", "-n", PROTO_DIR]
    print("********* CALLING 2to3 ***********")
    print(" ".join(two_to_three_command))
    if subprocess.call(two_to_three_command) != 0:
        sys.exit(-1)

    return True

if __name__ == '__main__':
    build_bayesmix(4)

File: R/bayesmixr/R/build_bayesmix.R
#' Builds the BayesMix executable
#'
#' After the build, if no error has occurred, it saves the path into the \code{BAYESMIX_EXE} environment variable.
#' Such variable is defined only when this package is loaded in the R session.
#'
#' @param nproc Number of processes to use for parallel compilation. Thanks to \code{parallel} package,
#' this parameter defaults to half of the available processes (through \code{\link[parallel]{detectCores}} function)
#' @param build_subdir Name for the sub-directory of \code{bayesmix/} folder in which configuration and compilation happens.
#' Default value is \code{build}.
#' @return No output if build is successful, it raises errors otherwise
#'
#' @export
build_bayesmix <- function(nproc = ceiling(parallel::detectCores()/2), build_subdir = "build") {

  # Check input types
  if(!is.numeric(nproc)) { stop("nproc must be a number") }
  if(!is.character(build_subdir)) { stop("build_subdir must be a string") }

  # Get .Renviron file from package
  renviron = system.file("bayesmixr.Renviron", package = "bayesmixr")

  # Set bayesmix_home folder from BAYESMIXR_HOME
  readRenviron(renviron)
  home_dir = Sys.getenv("BAYESMIXR_HOME")
  if(home_dir == ""){
    stop("Something went wrong while installing bayesmixr R package.")
  }
  bayesmix_home = dirname(dirname(home_dir))

  # Create build/ subdirectory
  build_dir = sprintf("%s/%s", bayesmix_home, build_subdir)
  dir.create(build_dir, showWarnings = F)

  # Configure bayesmix
  cat("*** Configuring BayesMix ***\n")
  flags = '-DDISABLE_TESTS=TRUE -DDISABLE_PLOTS=TRUE -DCMAKE_BUILD_TYPE=Release'
  CONFIGURE = sprintf('cmake .. -G "Unix Makefiles" %s', flags)
  errlog <- withr::with_dir(build_dir, system(CONFIGURE, ignore.stderr = TRUE))
  if(errlog != 0L){
    errmsg <- "Something went wrong during configure: command '%s' exit with status %d"
    stop(sprintf(errmsg, CONFIGURE, errlog))
  }
  cat("\n")

  # Build bayesmix::run_mcmc executable
  cat("*** Building BayesMix executable ***\n")
  BUILD = sprintf('make run_mcmc -j%d', nproc)
  errlog <- withr::with_dir(build_dir, system(BUILD))
  if (errlog != 0L) {
    errmsg <- "Something went wrong during build: command '%s' exit with status %d"
    stop(sprintf(errmsg, BUILD, errlog))
  }
  cat("\n")

  # Set BAYESMIX_EXE environment variable
  cat("*** Setting BAYESMIX_EXE environment variable ***\n")
  write(x = sprintf('BAYESMIX_EXE=%s/run_mcmc', build_dir), file = renviron, append = TRUE)
  cat("\n")

  # Set TBB_PATH environment variable
  cat("*** Setting TBB_PATH environment variable ***\n")
  tbb_path = sprintf('%s/lib/_deps/math-src/lib/tbb', bayesmix_home)
  write(x = sprintf('TBB_PATH=%s', tbb_path), file = renviron, append = TRUE)
  cat("\n")

  # Parse .Renviron file to get environment variables
  readRenviron(renviron)
  cat("Successfully installed BayesMix\n")
}
